lib/evmone/eof.cpp

```diff
@@ -34,51 +61,85 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
-                    return EOFValidationError::code_section_missing;
                 state = State::terminated;
                 break;
-            case DATA_SECTION:
-                if (section_headers[CODE_SECTION] == 0)
-                    return EOFValidationError::code_section_missing;
-                if (section_headers[DATA_SECTION] != 0)
-                    return EOFValidationError::multiple_data_sections;
+            case TYPE_SECTION:
+                expected_section_id = CODE_SECTION;
                 state = State::section_size;
                 break;
             case CODE_SECTION:
-                if (section_headers[CODE_SECTION] != 0)
-                    return EOFValidationError::multiple_code_sections;
+            {
+                if (it >= container_end - 2)
+                    return EOFValidationError::incomplete_section_number;
+                section_num = read_uint16_be(it);
+                it += 2;
+                if (section_num == 0)
+                    return EOFValidationError::zero_section_size;
+                expected_section_id = DATA_SECTION;
+                state = State::section_size;
+                break;
+            }
+            case DATA_SECTION:
+                expected_section_id = TERMINATOR;
                 state = State::section_size;
                 break;
             default:
-                return EOFValidationError::unknown_section_id;
+                assert(false);
             }
             break;
         }
         case State::section_size:
         {
-            const auto size_hi = *it++;
-            if (it == container_end)
-                return EOFValidationError::incomplete_section_size;
-            const auto size_lo = *it++;
-            const auto section_size = static_cast<uint16_t>((size_hi << 8) | size_lo);
-            if (section_size == 0)
-                return EOFValidationError::zero_section_size;
-
-            section_headers[section_id] = section_size;
+            if (section_id == CODE_SECTION)
+            {
+                assert(section_num > 0);  // Guaranteed by previous validation step.
+                for (size_t i = 0; i < section_num; ++i)
+                {
+                    if (it >= container_end - 2)
+                        return EOFValidationError::incomplete_section_size;
+                    const auto section_size = read_uint16_be(it);
+                    it += 2;
+                    if (section_size == 0)
+                        return EOFValidationError::zero_section_size;
+
+                    if (section_headers[CODE_SECTION].size() == CODE_SECTION_NUMBER_LIMIT)
+                        return EOFValidationError::too_many_code_sections;
+                    section_headers[section_id].emplace_back(section_size);
+                }
+            }
+            else  // TYPES_SECTION or DATA_SECTION
+            {
+                if (it + 1 == container_end)
```

Comment:
this will not detect the case when both bytes of section_size are truncated
```suggestion
                if (it + 1 >= container_end)
```

Reply:
The while loop detects the other part, but I changed it for consistency.